File: codes/ruby/chapter10/answer_A77.rb
# 入力
N, L = gets.split.map(&:to_i)
K = gets.to_i
A = gets.split.map(&:to_i)

# スコアの最大値が x 以上かを判定する関数
def check(x, n, l, k, a)
  cnt = 0         # 現時点で何回切ったかを表す
  last_kireme = 0 # 最後どこで切ったかを表す
  (0...n).each do |i|
    if a[i] - last_kireme >= x && l - a[i] >= x
      cnt += 1
      last_kireme = a[i]
    end
  end
  cnt >= k
end

# 二分探索（left: 現在の上限 ／ right: 現在の下限）
left, right = 1, 10**9
while left < right
  mid = (left + right + 1) / 2
  answer = check(mid, N, L, K, A)
  if !answer
    right = mid - 1 # 答えが前半部分に絞られる
  else
    left = mid      # 答えが後半部分に絞られる
  end
end

# 出力
puts left

File: codes/go/chapter10/answer_A77.go
package main

import (
	"fmt"
)

// check 関数: スコアの最大値が x 以上かを判定
func check(x int, n int, l int, k int, a []int) bool {
	cnt := 0
	lastKireme := 0
	for i := 0; i < n; i++ {
		if a[i]-lastKireme >= x && l-a[i] >= x {
			cnt++
			lastKireme = a[i]
		}
	}
	return cnt >= k
}

func main() {
	var n, l, k int
	fmt.Scan(&n, &l)
	fmt.Scan(&k)

	a := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i])
	}

	left, right := 1, 1000000000
	for left < right {
		mid := (left + right + 1) / 2
		answer := check(mid, n, l, k, a)
		if !answer {
			right = mid - 1
		} else {
			left = mid
		}
	}

	fmt.Println(left)
}
